File: sdk/ts/src/brokers/pubsub.ts
import { DEFAULT_CONFIG } from '../config';
import { SlowConsumerError } from '../errors';
import { NexoConnection } from '../transport/tcp/connection';
import { Logger } from '../utils/logger';
import { Subscription } from '../subscription';
import { FLAG_PUBSUB_PUB_CLEAR, FLAG_PUBSUB_PUB_HAS_TTL, FLAG_PUBSUB_PUB_RETAIN, PubSubOpcode } from '../protocol/generated';

const PubSubCommands = {
  publish: (conn: NexoConnection, topic: string, data: any, options: PublishOptions) => {
    const retain = options.retain === true;
    if (options.ttl !== undefined && (options.ttl < 0 || !Number.isInteger(options.ttl))) {
      throw new Error(`[PubSub] Invalid ttl: ${options.ttl}`);
    }
    const hasTtl = options.ttl !== undefined;
    const flags = (retain ? FLAG_PUBSUB_PUB_RETAIN : 0x00) | (hasTtl ? FLAG_PUBSUB_PUB_HAS_TTL : 0x00);
    return conn.send(PubSubOpcode.PUB, w => {
      w.string(topic).u8(flags);
      if (hasTtl) w.u32(options.ttl!);
      w.any(data);
    });
  },

  clear: (conn: NexoConnection, topic: string) =>
    conn.send(PubSubOpcode.PUB, w => w.string(topic).u8(FLAG_PUBSUB_PUB_CLEAR).any(Buffer.alloc(0))),

  subscribe: (conn: NexoConnection, pattern: string) =>
    conn.send(PubSubOpcode.SUB, w => w.string(pattern)),

  unsubscribe: (conn: NexoConnection, pattern: string) =>
    conn.send(PubSubOpcode.UNSUB, w => w.string(pattern)),
};

export interface PublishOptions {
  retain?: boolean;
  ttl?: number;
}

export interface PubSubSubscribeOptions {
  queueCapacity?: number;
}

export interface PubSubMessageMeta {
  topic: string;
}

export type PubSubHandler<T> = (data: T, meta: PubSubMessageMeta) => unknown | Promise<unknown>;

interface PatternEntry {
  readonly pattern: string;
  readonly wildcard: boolean;
  readonly listeners: Map<number, LocalListener<any>>;
  wireSubscribed: boolean;
  transition: Promise<void>;
}

class TrieNode {
  readonly literals = new Map<string, TrieNode>();
  plus: TrieNode | null = null;
  terminal: PatternEntry | null = null;
  hash: PatternEntry | null = null;
}

class PatternTrie {
  private readonly root = new TrieNode();

  insert(pattern: string, entry: PatternEntry): void {
    const parts = pattern.split('/');
    let node = this.root;
    for (const part of parts) {
      if (part === '#') {
        node.hash = entry;
        return;
      }
      if (part === '+') {
        if (node.plus === null) node.plus = new TrieNode();
        node = node.plus;
      } else {
        let child = node.literals.get(part);
        if (!child) {
          child = new TrieNode();
          node.literals.set(part, child);
        }
        node = child;
      }
    }
    node.terminal = entry;
  }

  remove(pattern: string, entry: PatternEntry): void {
    const parts = pattern.split('/');
    const path: { parent: TrieNode; part: string; node: TrieNode }[] = [];
    let node = this.root;
    for (const part of parts) {
      if (part === '#') {
        if (node.hash === entry) node.hash = null;
        this.prune(path);
        return;
      }
      const child = part === '+' ? node.plus : node.literals.get(part);
      if (!child) return;
      path.push({ parent: node, part, node: child });
      node = child;
    }
    if (node.terminal === entry) node.terminal = null;
    this.prune(path);
  }

  match(topic: string, visit: (entry: PatternEntry) => void): void {
    const parts = topic.split('/');
    const stack: { node: TrieNode; depth: number }[] = [{ node: this.root, depth: 0 }];
    while (stack.length > 0) {
      const current = stack.pop()!;
      if (current.node.hash) visit(current.node.hash);
      if (current.depth === parts.length) {
        if (current.node.terminal) visit(current.node.terminal);
        continue;
      }
      const literal = current.node.literals.get(parts[current.depth]);
      if (literal) stack.push({ node: literal, depth: current.depth + 1 });
      if (current.node.plus) stack.push({ node: current.node.plus, depth: current.depth + 1 });
    }
  }

  private prune(path: { parent: TrieNode; part: string; node: TrieNode }[]): void {
    for (let index = path.length - 1; index >= 0; index--) {
      const { parent, part, node } = path[index];
      if (node.terminal || node.hash || node.plus || node.literals.size > 0) return;
      if (part === '+') parent.plus = null;
      else parent.literals.delete(part);
    }
  }
}

class LocalListener<T> {
  active = true;
  failure: SlowConsumerError | null = null;
  private readonly queue: Array<{ data: T; meta: PubSubMessageMeta } | undefined> = [];
  private head = 0;
  private tail = 0;
  private size = 0;
  private running = false;
  private resolveClosed!: () => void;
  readonly closed: Promise<void>;

  constructor(
    private readonly handler: PubSubHandler<T>,
    private readonly capacity: number,
    private readonly logger: Logger,
    private readonly onOverflow: (error: SlowConsumerError) => void,
  ) {
    this.closed = new Promise(resolve => {
      this.resolveClosed = resolve;
    });
  }

  enqueue(data: T, meta: PubSubMessageMeta): void {
    if (!this.active) return;
    if (this.size === this.capacity) {
      const error = new SlowConsumerError(`PubSub listener queue exceeded capacity ${this.capacity}`);
      this.failure = error;
      this.deactivate();
      this.onOverflow(error);
      return;
    }

    this.queue[this.tail] = { data, meta };
    this.tail = (this.tail + 1) % this.capacity;
    this.size++;
    if (!this.running) {
      this.running = true;
      void this.consume();
    }
  }

  deactivate(): void {
    if (!this.active) return;
    this.active = false;
    this.queue.length = 0;
    this.head = 0;
    this.tail = 0;
    this.size = 0;
    if (!this.running) this.resolveClosed();
  }

  private async consume(): Promise<void> {
    while (this.active && this.size > 0) {
      const message = this.queue[this.head]!;
      this.queue[this.head] = undefined;
      this.head = (this.head + 1) % this.capacity;
      this.size--;
      try {
        await this.handler(message.data, message.meta);
      } catch (error) {
        this.logger.error('[PubSub] handler error', error);
      }
    }
    this.running = false;
    if (!this.active) this.resolveClosed();
  }
}

export class NexoTopic<T = any> {
  constructor(private readonly broker: NexoPubSub, public readonly name: string) { }

  publish(data: T, options: PublishOptions = {}): Promise<void> {
    return this.broker.publish(this.name, data, options);
  }

  clearRetained(): Promise<void> {
    return this.broker.clearRetained(this.name);
  }

  subscribe(callback: PubSubHandler<T>, options: PubSubSubscribeOptions = {}): Promise<Subscription> {
    return this.broker.subscribe(this.name, callback, options);
  }
}

export class NexoPattern<T = any> {
  constructor(private readonly broker: NexoPubSub, public readonly pattern: string) { }

  subscribe(callback: PubSubHandler<T>, options: PubSubSubscribeOptions = {}): Promise<Subscription> {
    return this.broker.subscribe(this.pattern, callback, options);
  }
}

export class NexoPubSub {
  private readonly entries = new Map<string, PatternEntry>();
  private readonly exact = new Map<string, PatternEntry>();
  private readonly wildcard = new PatternTrie();
  private nextListenerId = 1;

  constructor(private readonly conn: NexoConnection, private readonly logger: Logger) {
    conn.onPush = (topic, data) => this.dispatch(topic, data);

    conn.on('reconnect', () => {
      void this.restoreSubscriptions();
    });
  }

  topic<T = any>(concrete: string): NexoTopic<T> {
    NexoPubSub.validateTopic(concrete);
    return new NexoTopic<T>(this, concrete);
  }

  pattern<T = any>(wildcard: string): NexoPattern<T> {
    NexoPubSub.validatePattern(wildcard);
    return new NexoPattern<T>(this, wildcard);
  }

  async publish(topic: string, data: any, options: PublishOptions): Promise<void> {
    await PubSubCommands.publish(this.conn, topic, data, options);
  }

  async clearRetained(topic: string): Promise<void> {
    await PubSubCommands.clear(this.conn, topic);
  }

  async subscribe<T>(pattern: string, callback: PubSubHandler<T>, options: PubSubSubscribeOptions): Promise<Subscription> {
    const capacity = options.queueCapacity ?? DEFAULT_CONFIG.pubsub.listenerQueueCapacity;
    if (!Number.isInteger(capacity) || capacity < 1) {
      throw new Error(`queueCapacity must be a positive integer, got ${capacity}`);
    }

    let entry = this.entries.get(pattern);
    if (!entry) {
      entry = {
        pattern,
        wildcard: NexoPubSub.isWildcard(pattern),
        listeners: new Map(),
        wireSubscribed: false,
        transition: Promise.resolve(),
      };
      this.entries.set(pattern, entry);
      if (entry.wildcard) this.wildcard.insert(pattern, entry);
      else this.exact.set(pattern, entry);
    }

    const listenerId = this.nextListenerId++;
    const listener = new LocalListener<T>(callback, capacity, this.logger, error => {
      void this.handleOverflow(entry!, listenerId, listener, error);
    });
    entry.listeners.set(listenerId, listener);

    try {
      await this.ensureWireSubscribed(entry);
      if (listener.failure) throw listener.failure;
    } catch (error) {
      await this.detach(entry, listenerId, listener).catch(() => undefined);
      throw error;
    }

    return new Subscription(
      () => this.detach(entry!, listenerId, listener),
      () => listener.active,
      listener.closed,
      () => listener.failure,
    );
  }

  private dispatch(topic: string, data: any): void {
    const exactEntry = this.exact.get(topic);
    if (exactEntry) this.dispatchEntry(exactEntry, topic, data);
    this.wildcard.match(topic, entry => this.dispatchEntry(entry, topic, data));
  }

  private dispatchEntry(entry: PatternEntry, topic: string, data: any): void {
    const meta = { topic };
    for (const listener of entry.listeners.values()) {
      listener.enqueue(data, meta);
    }
  }

  private async detach(entry: PatternEntry, listenerId: number, listener: LocalListener<any>): Promise<void> {
    if (entry.listeners.get(listenerId) !== listener) {
      await listener.closed;
      return;
    }

    entry.listeners.delete(listenerId);
    listener.deactivate();
    let transition: Promise<void> = Promise.resolve();
    if (entry.listeners.size === 0) {
      transition = this.enqueueTransition(entry, async () => {
        if (entry.listeners.size > 0) return;
        try {
          if (entry.wireSubscribed) await PubSubCommands.unsubscribe(this.conn, entry.pattern);
        } finally {
          entry.wireSubscribed = false;
          if (entry.listeners.size === 0) this.removeEntry(entry);
        }
      });
    }
    await Promise.all([transition, listener.closed]);
  }

  private ensureWireSubscribed(entry: PatternEntry): Promise<void> {
    return this.enqueueTransition(entry, async () => {
      if (entry.listeners.size === 0 || entry.wireSubscribed) return;
      await PubSubCommands.subscribe(this.conn, entry.pattern);
      entry.wireSubscribed = true;
    });
  }

  private enqueueTransition(entry: PatternEntry, operation: () => Promise<void>): Promise<void> {
    const transition = entry.transition.catch(() => undefined).then(operation);
    entry.transition = transition;
    void transition.catch(() => undefined);
    return transition;
  }

  private removeEntry(entry: PatternEntry): void {
    if (this.entries.get(entry.pattern) !== entry || entry.listeners.size > 0) return;
    this.entries.delete(entry.pattern);
    if (entry.wildcard) this.wildcard.remove(entry.pattern, entry);
    else this.exact.delete(entry.pattern);
  }

  private async handleOverflow(entry: PatternEntry, listenerId: number, listener: LocalListener<any>, error: SlowConsumerError): Promise<void> {
    this.logger.error(`[PubSub] ${error.message}`);
    try {
      await this.detach(entry, listenerId, listener);
    } catch (detachError) {
      this.logger.error(`[PubSub] Failed to stop slow listener for "${entry.pattern}"`, detachError);
    }
  }

  private async restoreSubscriptions(): Promise<void> {
    const entries = Array.from(this.entries.values()).filter(entry => entry.listeners.size > 0);
    if (entries.length === 0) return;
    this.logger.info(`[PubSub] Restoring ${entries.length} subscription(s)...`);
    const results = await Promise.allSettled(entries.map(entry => this.enqueueTransition(entry, async () => {
      entry.wireSubscribed = false;
      if (entry.listeners.size === 0) return;
      await PubSubCommands.subscribe(this.conn, entry.pattern);
      entry.wireSubscribed = true;
    })));
    results.forEach((result, index) => {
      if (result.status === 'rejected') {
        this.logger.error(`[PubSub] Failed to resubscribe to ${entries[index].pattern}`, result.reason);
      }
    });
  }

  private static isWildcard(pattern: string): boolean {
    return pattern.split('/').some(part => part === '+' || part === '#');
  }

  private static validateTopic(topic: string): void {
    const parts = topic.split('/');
    if (!topic || parts.some(part => part === '')) throw new Error('[PubSub] Topic cannot contain empty segments');
    if (parts.some(part => part.includes('+') || part.includes('#'))) {
      throw new Error(`[PubSub] Concrete topic cannot contain wildcards: "${topic}"`);
    }
  }

  private static validatePattern(pattern: string): void {
    const parts = pattern.split('/');
    if (!pattern || parts.some(part => part === '')) throw new Error('[PubSub] Pattern cannot contain empty segments');
    let wildcard = false;
    for (let index = 0; index < parts.length; index++) {
      const part = parts[index];
      if (part.includes('+') && part !== '+') throw new Error('[PubSub] + wildcard must occupy an entire segment');
      if (part.includes('#') && part !== '#') throw new Error('[PubSub] # wildcard must occupy an entire segment');
      if (part === '#' && index !== parts.length - 1) throw new Error('[PubSub] # wildcard must be the last segment');
      wildcard ||= part === '+' || part === '#';
    }
    if (!wildcard) throw new Error(`[PubSub] Pattern must contain a wildcard segment: "${pattern}"`);
  }
}

File: sdk/py/src/nexo/brokers/pubsub.py
from __future__ import annotations

import asyncio
import inspect
import itertools
from typing import Any, Callable, Generic, TypeVar, TypedDict

from ..config import DEFAULT_CONFIG
from ..errors import SlowConsumerError
from ..protocol.generated import (
    FLAG_PUBSUB_PUB_CLEAR,
    FLAG_PUBSUB_PUB_HAS_TTL,
    FLAG_PUBSUB_PUB_RETAIN,
    PubSubOpcode,
)
from ..subscription import Subscription
from ..transport.tcp.connection import NexoConnection
from ..utils.logger import Logger


T = TypeVar("T")


class PubSubMessageMeta(TypedDict):
    topic: str


PubSubHandler = (
    Callable[[T], Any]
    | Callable[[T, PubSubMessageMeta], Any]
)


def _callback_accepts_meta(callback: Callable[..., Any]) -> bool:
    try:
        signature = inspect.signature(callback)
    except (TypeError, ValueError):
        return True
    positional = sum(
        parameter.kind in (parameter.POSITIONAL_ONLY, parameter.POSITIONAL_OR_KEYWORD)
        for parameter in signature.parameters.values()
    )
    return positional >= 2 or any(
        parameter.kind == parameter.VAR_POSITIONAL
        for parameter in signature.parameters.values()
    )


class _Listener:
    __slots__ = (
        "id",
        "handler",
        "wants_meta",
        "queue",
        "task",
        "active",
        "processing",
        "failure",
    )

    def __init__(self, listener_id: int, handler: Callable[..., Any], capacity: int) -> None:
        self.id = listener_id
        self.handler = handler
        self.wants_meta = _callback_accepts_meta(handler)
        self.queue: asyncio.Queue[tuple[Any, PubSubMessageMeta]] = asyncio.Queue(
            maxsize=capacity
        )
        self.task: asyncio.Task[None] | None = None
        self.active = True
        self.processing = False
        self.failure: SlowConsumerError | None = None


class _PatternState:
    __slots__ = ("pattern", "wildcard", "listeners")

    def __init__(self, pattern: str, wildcard: bool) -> None:
        self.pattern = pattern
        self.wildcard = wildcard
        self.listeners: dict[int, _Listener] = {}


class _MatcherNode:
    __slots__ = ("children", "single", "terminal", "multi")

    def __init__(self) -> None:
        self.children: dict[str, _MatcherNode] = {}
        self.single: _MatcherNode | None = None
        self.terminal: _PatternState | None = None
        self.multi: _PatternState | None = None


class _PatternMatcher:
    def __init__(self) -> None:
        self._root = _MatcherNode()

    def add(self, state: _PatternState) -> None:
        node = self._root
        for segment in state.pattern.split("/"):
            if segment == "#":
                node.multi = state
                return
            if segment == "+":
                if node.single is None:
                    node.single = _MatcherNode()
                node = node.single
            else:
                node = node.children.setdefault(segment, _MatcherNode())
        node.terminal = state

    def remove(self, state: _PatternState) -> None:
        segments = state.pattern.split("/")

        def remove_from(node: _MatcherNode, index: int) -> bool:
            segment = segments[index]
            if segment == "#":
                if node.multi is state:
                    node.multi = None
            elif segment == "+":
                child = node.single
                if child is not None and index + 1 < len(segments):
                    if remove_from(child, index + 1):
                        node.single = None
                elif child is not None:
                    if child.terminal is state:
                        child.terminal = None
                    if self._empty(child):
                        node.single = None
            else:
                child = node.children.get(segment)
                if child is not None and index + 1 < len(segments):
                    if remove_from(child, index + 1):
                        node.children.pop(segment, None)
                elif child is not None:
                    if child.terminal is state:
                        child.terminal = None
                    if self._empty(child):
                        node.children.pop(segment, None)
            return self._empty(node)

        remove_from(self._root, 0)

    def match(self, topic: str) -> list[_PatternState]:
        frontier = [self._root]
        matches: list[_PatternState] = []
        for segment in topic.split("/"):
            next_frontier: list[_MatcherNode] = []
            for node in frontier:
                if node.multi is not None:
                    matches.append(node.multi)
                literal = node.children.get(segment)
                if literal is not None:
                    next_frontier.append(literal)
                if node.single is not None:
                    next_frontier.append(node.single)
            frontier = next_frontier
            if not frontier:
                break
        for node in frontier:
            if node.terminal is not None:
                matches.append(node.terminal)
            if node.multi is not None:
                matches.append(node.multi)
        return matches

    @staticmethod
    def _empty(node: _MatcherNode) -> bool:
        return (
            not node.children
            and node.single is None
            and node.terminal is None
            and node.multi is None
        )


def _validate_segments(value: str) -> list[str]:
    if not value:
        raise ValueError("PubSub topic cannot be empty")
    segments = value.split("/")
    if any(not segment for segment in segments):
        raise ValueError("PubSub topic cannot contain empty segments")
    return segments


def _validate_topic(topic: str) -> None:
    segments = _validate_segments(topic)
    if any("+" in segment or "#" in segment for segment in segments):
        raise ValueError("Concrete PubSub topic cannot contain wildcards")


def _validate_pattern(pattern: str) -> None:
    segments = _validate_segments(pattern)
    wildcard = False
    for index, segment in enumerate(segments):
        if "+" in segment and segment != "+":
            raise ValueError("+ wildcard must occupy an entire segment")
        if "#" in segment and segment != "#":
            raise ValueError("# wildcard must occupy an entire segment")
        if segment == "#" and index != len(segments) - 1:
            raise ValueError("# wildcard must be the last segment")
        wildcard = wildcard or segment in {"+", "#"}
    if not wildcard:
        raise ValueError("PubSub pattern must contain a wildcard")


class NexoTopic(Generic[T]):
    def __init__(self, broker: "NexoPubSub", name: str) -> None:
        _validate_topic(name)
        self._broker = broker
        self.name = name

    async def publish(
        self,
        data: T,
        *,
        retain: bool = False,
        ttl: int | None = None,
    ) -> None:
        await self._broker._publish(
            self.name,
            data,
            retain=retain,
            ttl=ttl,
        )

    async def subscribe(
        self,
        callback: PubSubHandler[T],
        *,
        queue_capacity: int = DEFAULT_CONFIG.pubsub.listener_queue_capacity,
    ) -> Subscription[T]:
        return await self._broker._subscribe(
            self.name,
            callback,
            wildcard=False,
            queue_capacity=queue_capacity,
        )

    async def clear_retained(self) -> None:
        await self._broker._clear_retained(self.name)


class NexoPattern(Generic[T]):
    def __init__(self, broker: "NexoPubSub", pattern: str) -> None:
        _validate_pattern(pattern)
        self._broker = broker
        self.pattern = pattern

    async def subscribe(
        self,
        callback: PubSubHandler[T],
        *,
        queue_capacity: int = DEFAULT_CONFIG.pubsub.listener_queue_capacity,
    ) -> Subscription[T]:
        return await self._broker._subscribe(
            self.pattern,
            callback,
            wildcard=True,
            queue_capacity=queue_capacity,
        )


class NexoPubSub:
    def __init__(self, conn: NexoConnection, logger: Logger) -> None:
        self._conn = conn
        self._logger = logger
        self._states: dict[str, _PatternState] = {}
        self._exact: dict[str, _PatternState] = {}
        self._matcher = _PatternMatcher()
        self._listener_ids = itertools.count(1)
        self._registry_lock = asyncio.Lock()
        conn.on_push = self._enqueue
        conn.on_reconnect = self._restore_subscriptions

    def topic(self, name: str) -> NexoTopic[Any]:
        return NexoTopic(self, name)

    def pattern(self, wildcard: str) -> NexoPattern[Any]:
        return NexoPattern(self, wildcard)

    async def _publish(
        self,
        topic: str,
        data: Any,
        *,
        retain: bool,
        ttl: int | None,
    ) -> None:
        if ttl is not None and (
            not isinstance(ttl, int)
            or isinstance(ttl, bool)
            or ttl < 0
            or ttl > 0xFFFFFFFF
        ):
            raise ValueError(f"[PubSub] Invalid ttl: {ttl}")
        has_ttl = ttl is not None
        flags = (FLAG_PUBSUB_PUB_RETAIN if retain else 0x00) | (
            FLAG_PUBSUB_PUB_HAS_TTL if has_ttl else 0x00
        )

        def build(writer: Any) -> None:
            writer.string(topic).u8(flags)
            if has_ttl:
                writer.u32(ttl)
            writer.any(data)

        await self._conn.send(PubSubOpcode.PUB, build)

    async def _clear_retained(self, topic: str) -> None:
        await self._conn.send(
            PubSubOpcode.PUB,
            lambda writer: writer.string(topic)
            .u8(FLAG_PUBSUB_PUB_CLEAR)
            .any(b""),
        )

    async def _subscribe(
        self,
        pattern: str,
        callback: Callable[..., Any],
        *,
        wildcard: bool,
        queue_capacity: int,
    ) -> Subscription[Any]:
        if not isinstance(queue_capacity, int) or isinstance(queue_capacity, bool) or queue_capacity < 1:
            raise ValueError("queue_capacity must be a positive integer")
        async with self._registry_lock:
            state = self._states.get(pattern)
            first_listener = state is None
            if state is None:
                state = _PatternState(pattern, wildcard)
                self._register_state(state)
            elif state.wildcard != wildcard:
                raise ValueError("PubSub topic and pattern namespaces cannot overlap")

            listener = _Listener(next(self._listener_ids), callback, queue_capacity)
            state.listeners[listener.id] = listener
            listener.task = asyncio.create_task(self._consume(listener, pattern))
            if first_listener:
                try:
                    await self._conn.send(
                        PubSubOpcode.SUB,
                        lambda writer: writer.string(pattern),
                    )
                except BaseException:
                    state.listeners.pop(listener.id, None)
                    self._unregister_state(state)
                    await self._close_listener(listener)
                    raise

        async def stop_listener() -> None:
            await self._stop_listener(pattern, listener.id)

        return Subscription(
            stop_fn=stop_listener,
            active_fn=lambda: listener.active,
            completion=listener.task,
            error_fn=lambda: listener.failure,
        )

    async def _stop_listener(self, pattern: str, listener_id: int) -> None:
        listener: _Listener | None = None
        wire_error: Exception | None = None
        async with self._registry_lock:
            state = self._states.get(pattern)
            if state is None:
                return
            listener = state.listeners.pop(listener_id, None)
            if listener is None:
                return
            listener.active = False
            if not state.listeners:
                self._unregister_state(state)
                if self._conn.is_connected:
                    try:
                        await self._conn.send(
                            PubSubOpcode.UNSUB,
                            lambda writer: writer.string(pattern),
                        )
                    except Exception as error:
                        wire_error = error
        await self._close_listener(listener)
        if wire_error is not None:
            raise wire_error

    async def _close_listener(self, listener: _Listener) -> None:
        listener.active = False
        if listener.task is None:
            return
        if not listener.processing:
            listener.task.cancel()
        try:
            await listener.task
        except asyncio.CancelledError:
            pass

    async def _consume(self, listener: _Listener, pattern: str) -> None:
        try:
            while listener.active:
                data, meta = await listener.queue.get()
                if not listener.active:
                    break
                listener.processing = True
                try:
                    result = (
                        listener.handler(data, meta)
                        if listener.wants_meta
                        else listener.handler(data)
                    )
                    if inspect.isawaitable(result):
                        await result
                except Exception as error:
                    self._logger.error(
                        f"[PubSub:{pattern}] handler error: {error}"
                    )
                finally:
                    listener.processing = False
        except asyncio.CancelledError:
            pass
        finally:
            listener.active = False

    def _register_state(self, state: _PatternState) -> None:
        self._states[state.pattern] = state
        if state.wildcard:
            self._matcher.add(state)
        else:
            self._exact[state.pattern] = state

    def _unregister_state(self, state: _PatternState) -> None:
        self._states.pop(state.pattern, None)
        if state.wildcard:
            self._matcher.remove(state)
        else:
            self._exact.pop(state.pattern, None)

    def _enqueue(self, topic: str, data: Any) -> None:
        exact = self._exact.get(topic)
        if exact is not None:
            self._enqueue_state(exact, topic, data)
        for state in self._matcher.match(topic):
            self._enqueue_state(state, topic, data)

    def _enqueue_state(self, state: _PatternState, topic: str, data: Any) -> None:
        meta: PubSubMessageMeta = {"topic": topic}
        for listener in tuple(state.listeners.values()):
            if not listener.active:
                continue
            try:
                listener.queue.put_nowait((data, meta))
            except asyncio.QueueFull:
                listener.failure = SlowConsumerError(
                    f"PubSub listener queue exceeded capacity {listener.queue.maxsize}"
                )
                listener.active = False
                self._logger.error(f"[PubSub:{state.pattern}] {listener.failure}")
                asyncio.create_task(
                    self._stop_overflowed_listener(state.pattern, listener.id)
                )

    async def _stop_overflowed_listener(self, pattern: str, listener_id: int) -> None:
        try:
            await self._stop_listener(pattern, listener_id)
        except Exception as error:
            self._logger.error(
                f"[PubSub:{pattern}] Failed to stop slow listener: {error}"
            )

    async def _restore_subscriptions(self) -> None:
        async with self._registry_lock:
            patterns = tuple(self._states)
            if not patterns:
                return
            self._logger.info(
                f"[PubSub] Restoring {len(patterns)} subscription(s)..."
            )
            async def restore(pattern: str) -> None:
                await self._conn.send(
                    PubSubOpcode.SUB,
                    lambda writer: writer.string(pattern),
                )

            results = await asyncio.gather(
                *(restore(pattern) for pattern in patterns),
                return_exceptions=True,
            )
            for pattern, result in zip(patterns, results):
                if isinstance(result, BaseException):
                    self._logger.error(
                        f"[PubSub] Failed to resubscribe to {pattern}",
                        result,
                    )
